### ISSUE TITLE
Vertical tabs break ODS files

### ISSUE THREAD

**Issue by madflow** [box/spout#329]
\v breaks the ODS writer

```php
$writer = \Box\Spout\Writer\WriterFactory::create(Type::ODS);
$writer->openToFile(__DIR__ . '/failed.ods');
$writer->addRow(["\v"]);
$writer->close();
```

> Read-Error.
Format error discovered in the file in sub-document content.xml at 9,387(row,col)

Tested with LibreOffice 5.2.

CSV and XLSX work fine.

Should we fix this in upstream (spout)?

This came up from user input, probably copy-pasted from somewhere.

**Comment by adrilo** [comment]
Yes, it'd be nice to have a fix in Spout, so that other developers would benefit from it.
Feel free to submit a pull request!

BTW, is it broken with other characters too?

**Comment by madflow** [comment]
A vertical tab is a illegal character in this xml spec. 

http://stackoverflow.com/questions/1091945/what-characters-do-i-need-to-escape-in-xml-documents
https://www.w3.org/TR/xml/#charsets

I also found, that a form feed breaks the XML.

There already seems to be an escaping mechanism for XLSX content with some heavy preg_replace'ing - maybe we can adapt something form there.


```php
require __DIR__ . '/vendor/autoload.php';

use Box\Spout\Reader\ReaderFactory;
use Box\Spout\Common\Type;
use Box\Spout\Writer\WriterFactory;

$writerTypes = [Type::ODS, Type::XLSX];

$chars = ["\v", "\f", "<", ">", "&", ","];

foreach($writerTypes as $writerType) {

    foreach($chars as $index => $char) {

        $writer = WriterFactory::create($writerType);
        $fileName = sprintf('/invalid_chars_%s.%s', $writerType, $writerType);
        $writer->openToFile(__DIR__ . $fileName);
        $writer->addRow([$char]);
        $writer->close();

        $reader = ReaderFactory::create($writerType);
        $reader->open(__DIR__ . $fileName);

        try {
            /** @var \Box\Spout\Reader\XLSX\Sheet $sheet */
            foreach($reader->getSheetIterator() as $sheet) {
                foreach($sheet->getRowIterator() as $row) {

                }
            }
        } catch(\Exception $e) {
            echo sprintf('The char at index %s is invalid in writer %s', $index, $writerType).PHP_EOL;
        }

        $reader->close();

    }
}
```

> The char at index 0 is invalid in writer ods
> The char at index 1 is invalid in writer ods

**Comment by madflow** [comment]
I tried to fix this by using PHPs advice.

http://php.net/manual/en/function.htmlspecialchars.php

> ENT_DISALLOWED 	
> Replace invalid code points for the given document type with a Unicode Replacement Character U+FFFD (UTF-8) or &#xFFFD; (otherwise) instead of leaving them as is. This may be useful, for instance, to ensure the well-formedness of XML documents with embedded external content. 

However the HVM test fail, because the constant is not supported :S

**Comment by adrilo** [comment]
For reference: https://github.com/facebook/hhvm/issues/4938

**Comment by adrilo** [comment]
So basically, adding this to \Box\Spout\Common\Escaper\ODS.php:

```php
defined('ENT_DISALLOWED') || define('ENT_DISALLOWED', 128);
```

Can you try if that fixes the test?

**Comment by madflow** [comment]
This fixes the constant warning - but the tests fail. Is HVM a "first class citizen"? 

+ Can we mark hvm as "allow_failures" in travis?
+ We could adjust the test to be skipped on hvm. 
+ Hack around it/another solution?

**Comment by adrilo** [comment]
Allowing test failures seems like a bad pattern. Skipping the test and giving a reason would be a better approach.

Now, this error is most likely really rare. Using ENT_DISALLOWED would make it even rarer, as only HHVM users using invalid characters would see it. It could be a good enough approach, as it's really simple. My only concern would be that there is no ETA for this constant to be supported in HHVM so we don't know how long the ODS writer will be broken for.

Another solution would consist in doing the same thing that the XLSX escaper does (and what ENT_DISALLOWED does). Replacing the invalid characters by �. The code would be something like this: `preg_replace('/[\x00-\x08\x0B-\x0C\x0E-\x1F]/', '�', $string);`

I guess I would vote for option 2, and add a note that this can be replaced by ENT_DISALLOWED as soon as it is available in HHVM.

**Comment by madflow** [comment]
I pushed "something"  :S  . I assumed your option 2 - not my option 2...

**Comment by adrilo** [comment]
Fixed in #333
